### PR TITLE
Cycle Web Scraper Script

### LINKED ISSUE #417: Beam Status Web-Scraper
Issue raised by: developer
### What?
A Web-Scraper to read the beam status from `https://www.isis.stfc.ac.uk/Pages/Beam-Status.aspx` into a pandas dataframe.

### Where?
scripts/system_performance/beam_status_webscraper.py

### How?
To acquire cycle dates for use in system health check MySQL queries

### How to test the issue is resolved

- Ensure all data from beam status cycle table is scraped and re-formatted appropriately for use in MySQL queries.
- The data frame is saved to a .csv file for use if a connection can’t be made when run again.
- All unit tests pass

### PR DESCRIPTION
**Pull request by JackEAllen**
### Summary of work
A web scraping script to retrieve cycle data from `https://www.isis.stfc.ac.uk/Pages/Beam-Status.aspx` for use in MySQL queries to monitor Autoreductions system performance

### How to test your work
- Test script runs with no errors both with the correct url passed and then with a fake url to make sure the script is able to retrieve a local copy of cycle dates placed in a .csv file. 
- Add `print(DATA_FRAME)` to the end of the script and run from your terminal to ensure that the start and end columns have been modified accordingly. 
- Check all unit tests pass

Fixes #417

### CODE REVIEW

File: scripts/system_performance/cycle_webscraper.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+
+class TableWebScraper:
+    """
+    Web-scrapes beam cycle data from https://www.isis.stfc.ac.uk/Pages/Beam-Status.aspx
+    for MySQL system checks. This URl can be changed at the bottom of the script
+    """
+
+    def __init__(self, url, local_data=None):
+        self.url = url
+        if local_data is None:
```
Comment (ElliotAOram) [inline scripts/system_performance/cycle_webscraper.py]: `if not local_data:` is more pythonic

File: scripts/system_performance/cycle_webscraper.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        else:
+            # Directory doesn't exist
+            logging.error("No file named {} found".format(local_data))
+
+    # -/////////////////////////////////////// Web scraping //////////////////////////////////-#
+
+    def create_table(self):
+        """Private function to format data scraped ready to be placed into a pandas data frame"""
+
+        def _create_header_cols(elements, column):
```
Comment (ElliotAOram) [inline scripts/system_performance/cycle_webscraper.py]: Would be beneficial to have documentation here: 
"""
:param elements: description of elements
:param column: blah 
"""

File: scripts/system_performance/cycle_webscraper.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+            # For each row, store each first element (header) and an empty list col
+            index = 0
```
Comment (ElliotAOram) [inline scripts/system_performance/cycle_webscraper.py]: what is index used for here?

Comment (JackEAllen) [reply scripts/system_performance/cycle_webscraper.py]: From a previous iterator design. Will be removed in next push

File: scripts/system_performance/cycle_webscraper.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+    def save_to_csv(self, data):
+        """ Create local copy of dataframe as .csv to read when web page is not accessible"""
+        data.to_csv(r'{}'.format(self.local_data), encoding='utf-8')
+        return data
+
+    @staticmethod
+    def read_csv(local_data):
+        """Reads csv into a pandas data frame """
+        if os.path.isfile("{}".format(local_data)):
```
Comment (ElliotAOram) [inline scripts/system_performance/cycle_webscraper.py]: What type is local_data here? I would suggest that perhaps `os.path.isfile(str(local_data)):` would be cleaner

File: scripts/system_performance/cycle_webscraper.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+        def _populate_table(elements, col):
+            """Populate  table with web scraped data"""
+
+            # Data is stored on the second row onwards
+            for j in range(1, len(elements)):
+                # tr_elem is our j'th row
+                tr_elem = elements[j]
+
+                # If row is not of size 5, the //tr data is not from table we expect
+                if len(tr_elem) != 5:
```
Comment (ElliotAOram) [inline scripts/system_performance/cycle_webscraper.py]: if the //tr data is not from the table we expect, is tr_elem still valid?

File: scripts/system_performance/cycle_webscraper.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+                    break
+
+                # Iterate through each element of the row
+                index = 0
+                for text in tr_elem.iterchildren():
```
Comment (ElliotAOram) [inline scripts/system_performance/cycle_webscraper.py]: you can simplify this with using enumerate: https://stackoverflow.com/questions/522563/accessing-the-index-in-for-loops

File: scripts/system_performance/cycle_webscraper.py [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+            doc = lh.fromstring(response.content)  # Store the contents of the website under doc
+            tr_elements = doc.xpath('//tr')  # Parse data stored between <tr>..</tr> of HTML
+            _create_header_cols(tr_elements, col)
+            _populate_table(tr_elements, col)
+            data_table = _datatable_constructor(col)
+            return data_table
+
+        def _check_connection(url):
+            """Web scrape if URL can be pinged else retrieve local copy made"""
+            request = requests.get('{}'.format(url))
```
Comment (ElliotAOram) [inline scripts/system_performance/cycle_webscraper.py]: `str(url)`

**Pull request merged**